subgraphs/arbitrum-one-bridge/src/handlers/arbitrum/transferIn.ts

```diff
@@ -0,0 +1,83 @@
+import { Address, ethereum, log } from "@graphprotocol/graph-ts";
+import { BridgeConfig } from "../../sdk/protocols/bridge/config";
+import {
+  DepositFinalized,
+  TokenGateway,
+} from "../../../generated/ERC20Gateway/TokenGateway";
+import {
+  BridgePermissionType,
+  BridgePoolType,
+  CrosschainTokenType,
+} from "../../sdk/protocols/bridge/enums";
+import { Versions } from "../../versions";
+import { SDK } from "../../sdk/protocols/bridge";
+import { networkToChainID } from "../../sdk/protocols/bridge/chainIds";
+import { arbSideConf, Pricer, TokenInit } from "../../common/utils";
+import { Network } from "../../sdk/util/constants";
+
+export function handleTransferIn3pGateway(event: DepositFinalized): void {
+  log.error("[3p Gateway] We are in transferIn3pGateway", []);
+
+  const l1Token = new ethereum.EventParam("l1token", event.parameters[0].value);
+  const _from = new ethereum.EventParam("_from", event.parameters[1].value);
+  const _to = new ethereum.EventParam("_to", event.parameters[2].value);
+  const _amount = new ethereum.EventParam("_amount", event.parameters[3].value);
+
+  const params: ethereum.EventParam[] = [l1Token, _from, _to, _amount];
+  const depositFinalized = new DepositFinalized(event.address, event.logIndex, event.transactionLogIndex, event.logType, event.block, event.transaction, params, event.receipt);
+  handleTransferIn(depositFinalized);
+}
+
+export function handleTransferIn(event: DepositFinalized): void {
+  // -- SDK
+
+  const sdk = SDK.initialize(arbSideConf, new Pricer(), new TokenInit(), event);
+
+  // -- TOKENS
+
+  const gatewayContract = TokenGateway.bind(event.address);
+  const inputTokenAddress = event.params.l1Token;
+  let crossTokenAddress: Address;
+
+  const crossTokenAddressResult =
+    gatewayContract.try_calculateL2TokenAddress(inputTokenAddress);
+  if (crossTokenAddressResult.reverted) {
+    log.info("calculate cross token address call reverted", []);
+  } else {
+    crossTokenAddress = crossTokenAddressResult.value;
+  }
+
+  const crossToken = sdk.Tokens.getOrCreateCrosschainToken(
+    networkToChainID(Network.ARBITRUM_ONE),
+    crossTokenAddress!,
+    CrosschainTokenType.CANONICAL,
+    crossTokenAddress!
+  );
+
+  // -- POOL
+
+  const poolId = event.address;
```

Comment:
Does this address handle multiple ERC20 tokens? Or there is a contract per each token?

In case a single address manages multiple tokens, you’ll need to have them as separate pools in the schema. Since each Pool in the schema only handles one token. You would have to break down the real pool into multiple “virtual” ones.

Reply:
Ah, didn’t realize we want to approach it that way but makes sense. I’ll break the pools down. Is there a naming convention for the pools in such case? (`address-token` work?).

@jaimehgb  Slightly related, how should the protocols be setup? Right now, I’m create a single protocol (or rather the sdk is) on arbitrum and ethereum side. The address I’m using for the protocol is a generic address (because there isn’t a single factory contract or equivalent that we can use). Do these make sense?

Reply:
Answering this, although already outdated, but for context:

As long as each pool ID is unique, it is fine. We don’t rely on IDs for anything else, since it is not always possible to make them unique from the contract address alone.

Same for the protocol. If you have a factory contract from which all the rest come, you can use that one. If not, you can use the deployer address. But same, the ID is not used for anything other than uniquely identifying entities, so anything goes.